File: coupon_system/src/main/java/com/jb/coupon/service/AdminService.java
package com.jb.coupon.service;

import java.util.List;
import java.util.Optional;

import com.jb.coupon.exception.CompanyNameInUseException;
import com.jb.coupon.exception.EmailInUseException;
import com.jb.coupon.exception.InvalidDetailsException;
import com.jb.coupon.exception.NotExistException;
import com.jb.coupon.exception.NullDetailsException;
import com.jb.coupon.exception.SomethingWentWrong;
import com.jb.coupon.model.Company;
import com.jb.coupon.model.Coupon;
import com.jb.coupon.model.Customer;

public interface AdminService {

	/**
	 * Function that adds a new company with unique name and email
	 * 
	 * @param company
	 * @throws NullDetailsException
	 * @throws InvalidDetailsException
	 * @throws EmailInUseException
	 * @throws CompanyNameInUseException
	 */
	void addCompany(Company company) throws NullDetailsException, InvalidDetailsException,
			EmailInUseException, CompanyNameInUseException;

	/**
	 * Function which updates an existing company except for company ID and company
	 * name
	 * 
	 * @param company
	 * @throws NullDetailsException
	 * @throws InvalidDetailsException
	 * @throws NotExistException
	 * @throws EmailInUseException
	 */
	void updateCompany(Company company) throws NullDetailsException, InvalidDetailsException, NotExistException,
			EmailInUseException;

	/**
	 * Function that deletes an existing company, its coupons and the history of
	 * coupons purchase by customers
	 * 
	 * @param companyId
	 * @throws NullDetailsException
	 * @throws InvalidDetailsException
	 * @throws NotExistException
	 */
	void deleteCompany(Long companyId) throws NullDetailsException, InvalidDetailsException, NotExistException;

	/**
	 * Function which returns all companies
	 * 
	 * @return Optional<List<Company>>
	 * @throws NotExistException
	 */
	Optional<List<Company>> getAllCompanies() throws NotExistException;

	/**
	 * This function returns a specific company by company ID
	 * 
	 * @param companyId
	 * @return Optional<Company>
	 * @throws NullDetailsException
	 * @throws InvalidDetailsException
	 * @throws NotExistException
	 */
	Optional<Company> getOneCompany(Long companyId)
			throws NullDetailsException, InvalidDetailsException, NotExistException;
	
	/**
	 * Function that returns company coupons by company ID
	 * 
	 * @param companyId
	 * @return Optional<List<Coupon>>
	 * @throws NullDetailsException
	 * @throws InvalidDetailsException
	 * @throws SomethingWentWrong
	 */
	Optional<List<Coupon>> getCompanyCoupons(Long companyId)
			throws NullDetailsException, InvalidDetailsException, SomethingWentWrong;

	/**
	 * Function which adds a new customer and verifies that its email is unique
	 * 
	 * @param customer
	 * @throws NullDetailsException
	 * @throws InvalidDetailsException
	 * @throws EmailInUseException
	 */
	void addCustomer(Customer customer)
			throws NullDetailsException, InvalidDetailsException, EmailInUseException;

	/**
	 * Function that updates an existing customer except for its ID
	 * 
	 * @param Customer
	 * @throws NullDetailsException
	 * @throws InvalidDetailsException
	 * @throws EmailInUseException
	 * @throws NotExistException
	 */
	void updateCustomer(Customer customer) throws NullDetailsException, InvalidDetailsException,
			EmailInUseException, NotExistException;

	/**
	 * Function which deletes an existing customer and its coupons purchase history
	 * 
	 * @param customerId
	 * @throws NullDetailsException
	 * @throws InvalidDetailsException
	 * @throws NotExistException
	 */
	void deleteCustomer(Long customerId)
			throws NullDetailsException, InvalidDetailsException, NotExistException;

	/**
	 * Function that returns all customers
	 * 
	 * @return Optional<List<Customer>>
	 * @throws NotExistException
	 */
	Optional<List<Customer>> getAllCustomers() throws NotExistException;

	/**
	 * Function that returns one customer by customer ID
	 * 
	 * @param customerId
	 * @return Optional<Customer>
	 * @throws NullDetailsException
	 * @throws InvalidDetailsException
	 * @throws NotExistException
	 */
	Optional<Customer> getOneCustomer(Long customerId)
			throws NullDetailsException, InvalidDetailsException, NotExistException;
	
	/**
	 * Function that returns customer coupons by customer ID
	 * 
	 * @param customerId
	 * @return Optional<List<Coupon>>
	 * @throws NullDetailsException
	 * @throws InvalidDetailsException
	 * @throws SomethingWentWrong
	 */
	Optional<List<Coupon>> getCustomerCoupons(Long customerId)
			throws NullDetailsException, InvalidDetailsException, SomethingWentWrong;
}

File: coupons-website/src/share/service/AdminService.js
// import Api - axios HTTP communication handler
import Api from '../api/Api';

// import Company, Customer and Coupon object model classes
import Company from '../model/core.company';
import Customer from '../model/core.customer';
import Coupon from '../model/core.coupon';

// import CatchErrors Component
import CatchErrors from './CatchErrors';

class AdminService {

    dispatch;

    constructor(dispatch) {
        this.dispatch = dispatch;
    };

    getAllCompanies = async () => {
        console.log('AdminService.js/getAllCompanies');
        return Api.admin().getAllCompanies()
            .then(response => {
                console.log('AdminService.js/getAllCompanies/response: ', response);
                const companies = response.data.map(company =>
                    new Company(
                        company.id,
                        company.name,
                        company.email,
                        company.password
                        ));
                return { succeeded: true, companies };
            }).catch(err => {
                return CatchErrors(this.dispatch, err);
            });
    };

    getOneCompany = async (id) => {
        console.log('AdminService.js/getOneCompany');
        return Api.admin().getOneCompany(id)
            .then(response => {
                console.log('AdminService.js/getOneCompany/response: ', response);
                const company = response.data;
                return { succeeded: true, company };
            }).catch(err => {
                return CatchErrors(this.dispatch, err);
            });
    };

    addCompany = async (company) => {
        console.log('AdminService.js/addCompany');
        return Api.admin().addCompany(company)
            .then(response => {
                console.log('AdminService.js/addCompany/response: ', response);
                return { succeeded: true };
            }).catch(err => {
                return CatchErrors(this.dispatch, err);
            });
    };

    updateCompany = async (updatedCompany) => {
        console.log('AdminService.js/updateCompany');
        return Api.admin().updateCompany(updatedCompany)
            .then(response => {
                console.log('AdminService.js/updateCompany/response: ', response);
                return { succeeded: true };
            }).catch(err => {
                return CatchErrors(this.dispatch, err);
            });
    };

    deleteCompany = async (id) => {
        console.log('AdminService.js/deleteCompany/id', id);
        return Api.admin().deleteCompany(id)
            .then(response => {
                console.log('AdminService.js/deleteCompany/response: ', response);
                return { succeeded: true };
            }).catch(err => {
                return CatchErrors(this.dispatch, err);
            });
    };

    getAllCustomers = async () => {
        console.log('AdminService.js/getAllCustomers');
        return Api.admin().getAllCustomers()
            .then(response => {
                console.log('AdminService.js/getAllCustomers/response: ', response);
                const customers = response.data.map(customer =>
                    new Customer(
                        customer.id,
                        customer.firstName,
                        customer.lastName,
                        customer.email,
                        customer.password
                    ));
                return { succeeded: true, customers };
            }).catch(err => {
                return CatchErrors(this.dispatch, err);
            });
    };

    getOneCustomer = async (id) => {
        console.log('AdminService.js/getOneCustomer');
        return Api.admin().getOneCustomer(id)
            .then(response => {
                console.log('AdminService/ getOneCustomer/ response: ', response);
                const customer = response.data;
                return { succeeded: true, customer };
            }).catch(err => {
                return CatchErrors(this.dispatch, err);
            });
    };

    addCustomer = async (customer) => {
        console.log('AdminService.js/addCustomer');
        return Api.admin().addCustomer(customer)
            .then(response => {
                console.log('AdminService.js/addCustomer/response: ', response);
                return { succeeded: true };
            }).catch(err => {
                return CatchErrors(this.dispatch, err);
            });
    };

    updateCustomer = async (updatedCustomer) => {
        console.log('AdminService.js/updateCustomer');
        return Api.admin().updateCustomer(updatedCustomer)
            .then(response => {
                console.log('AdminService.js/updateCustomer/response: ', response);
                return { succeeded: true };
            }).catch(err => {
                return CatchErrors(this.dispatch, err);
            });
    };

    deleteCustomer = async (id) => {
        console.log('AdminService.js/deleteCustomer');
        return Api.admin().deleteCustomer(id)
            .then(response => {
                console.log('AdminService.js/deleteCustomer/response: ', response);
                return { succeeded: true };
            }).catch(err => {
                return CatchErrors(this.dispatch, err);
            });
    };

    getCompanyCoupons = async (id) => {
        console.log('AdminService.js/getCompanyCoupons, id: ', id);
        return Api.admin().getCompanyCoupons(id)
            .then(response => {
                console.log('AdminService.js/getCompanyCoupons/response: ', response);
                const coupons = response.data.map(coupon =>
                    new Coupon(
                        coupon.id,
                        coupon.title,
                        coupon.company,
                        coupon.category,
                        coupon.description,
                        coupon.startDate,
                        coupon.endDate,
                        coupon.amount,
                        coupon.price,
                        coupon.image
                    ));
                return { succeeded: true, coupons };
            }).catch(err => {
                return CatchErrors(this.dispatch, err);
            });
    };

    getCustomerCoupons = async (id) => {
        console.log('AdminService.js/getCustomerCoupons, id: ', id);
        return Api.admin().getCustomerCoupons(id)
            .then(response => {
                console.log('AdminService.js/getCustomerCoupons/response: ', response);
                const coupons = response.data.map(coupon =>
                    new Coupon(
                        coupon.id,
                        coupon.title,
                        coupon.company,
                        coupon.category,
                        coupon.description,
                        coupon.startDate,
                        coupon.endDate,
                        coupon.amount,
                        coupon.price,
                        coupon.image
                    ));
                return { succeeded: true, coupons };
            }).catch(err => {
                return CatchErrors(this.dispatch, err);
            });
    };
};

export default AdminService;